text/0000-external-doc-attribute.md

```diff
@@ -0,0 +1,210 @@
+mod example {
+    #[external_doc("docs/example.md", "mod")]
+    fn my_func() {
+      // Hidden implementation
+    }
+}
+```
+
+It should expand out to:
+
+```rust
+mod example {
+    //! # I'm an example
+    //! This is a markdown file that gets imported to Rust as a doc comment.
+    fn my_func() {
+      // Hidden implementation
+    }
+}
+```
+
+## Line numbers when errors occur
+As with all macros being expanded this brings up the question of line numbers
+and for documentation tests especially so, to keep things simple for the user
+the documentation should be treated separately from the code. Since the
+attribute only needs to be expanded with `rustdoc` or `cargo test`, it should be
+ignored by the compiler except for having the proper lines for error messages.
+
+For example if we have this:
+
+```rust
+#[external_doc("docs/example.md", "line")] // Line 1
+f my_func() {                              // Line 2
+  // Hidden implementation                 // Line 3
+}                                          // Line 4
+```
+
+Then we would have a syntax error on line 2, however the doc comment comes
+before that. In this case the compiler would ignore the attribute for expansion,
+but would say that the error occurs on line 2 rather than saying it is line 1 if
+the attribute is ignored. This makes it easy for the user to spot their error.
+This same behavior should be observed in the case of inline tests and those in
+the tests directory.
+
+If we have a documentation test failure the line number should be for the
+external doc file and the line number where it fails, rather than a line number
+from the code base itself. Having the numbers for the lines being used because
+they were inserted into the code for these scenarios would cause confusion and
+would obfuscate where errors occur, making it harder not easier for end users,
+making this feature useless if it creates ergonomic overhead like this.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+`#[external_doc("path", "type")]` is an extension of the current
+`#[doc = "doc"]` attribute by allowing documentation to exist outside of the
+source code. This isn't entirely hard to grasp if one is familiar with
+attributes but if not then this syntax vs a `///` or `//!` type of comment
+could cause confusion. By labeling the attribute as `external_doc`, having a
+clear path and type (either `line` or `mod`) then should, at the very least,
+provide context as to what's going on and where to find this file for inclusion.
+
+The acceptance of this proposal would minimally impact all levels of Rust users
+as it is something that provides convenience but is not a necessary thing to
+learn to use Rust. It should be taught to existing users by updating
+documentation to show it in use and to include in in the Rust Programming
+Language book to teach new users. Currently the newest version of The Rust
+Programming Language does not include doc comments as part of it. This would
```

Comment:
Ah see I thought it would have been under the comments section which is why I missed that.